### ISSUE TITLE
Adding L2-regularization

### ISSUE THREAD

**Issue by rotemmairon** [tflearn/tflearn#860]
I wish to apply L2-regularization to the weights of a network. the function add_weight_regularizer seems relevant, but I am not sure how to use it for my purpose. Instead, I define a custom, L2-regularized objective function (described below) according to the conventions in tflearn/objectives.py.

**My question** is whether there is a more simple, straightforward way using add_weight_regularizer or other TFlearn library functions.

This is what I tried:

```
# true iff the given variables is a weight variable
def is_weight(var):
    return var.name.split("/")[1][0] == "W"

# retrive all weight variable in the network
def retrieve_weights():
    trainable_vars = tflearn.variables.get_all_trainable_variable()
    return [var for var in trainable_vars if is_weight(var)]

# calculate the penalty on the norm of the weights and add to the loss
def regularize_loss(loss, weights, beta):
    beta = tf.constant(beta, dtype=tf.float32, name="lambda")
    for weight in weights:
        reg_param = tf.multiply(beta, tf.nn.l2_loss(weight))
        loss = tf.add(loss, reg_param, name=weight.name.split("/")[0])
    return loss

# given the parameter beta (amount of regularization), generate a 
# custom function for regularized categorical-cross-entropy loss
def l2reg_categorical_crossentropy(beta=0.0005):
    def loss_func(y_pred, y_true):
        with tf.name_scope("L2RegCrossentropy"):
            y_pred /= tf.reduce_sum(y_pred,
                                    reduction_indices=len(y_pred.get_shape()) - 1,
                                    keep_dims=True)

            # Manual computation of crossentropy
            y_pred = tf.clip_by_value(y_pred, tf.cast(_EPSILON, dtype=_FLOATX),
                                      tf.cast(1. - _EPSILON, dtype=_FLOATX))
            cross_entropy = - tf.reduce_sum(y_true * tf.log(y_pred),
                                            reduction_indices=len(y_pred.get_shape()) - 1)

            # THE ONLY DIFFERENCE TO categorical_crossentropy IN tflearn/objectives.py*
            weights = retrieve_weights()
            loss = tf.reduce_mean(cross_entropy)
            loss = regularize_loss(loss, weights, beta)
        return loss
    return loss_func
```

Finaly, the new custom loss function is used as follows:

```
loss = utils.l2reg_categorical_crossentropy(beta=0.001)
adam = Adam(learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=0.1, use_locking=False)
# Model 
network = ...
:
network = regression(network, optimizer=adam, loss=loss)
# Train
...
```

**Comment by mubastan** [comment]
TFLearn already has L2 regularization in the layers:
http://tflearn.org/layers/conv/
`tflearn.layers.conv.conv_2d (incoming, nb_filter, filter_size, strides=1, padding='same', activation='linear', bias=True, weights_init='uniform_scaling', bias_init='zeros', regularizer=None, weight_decay=0.001, trainable=True, restore=True, reuse=False, scope=None, name='Conv2D')`

http://tflearn.org/layers/core/

`tflearn.layers.core.fully_connected (incoming, n_units, activation='linear', bias=True, weights_init='truncated_normal', bias_init='zeros', regularizer=None, weight_decay=0.001, trainable=True, restore=True, reuse=False, scope=None, name='FullyConnected')`

**Comment by rotemmairon** [comment]
Thank you, @mubastan, you comment was helpful. Combining your pointing out of the parameter _regularizer_, and the short regularization example in [TFlearn's guide](http://tflearn.org/getting_started/), I understand that regularization is added when the model is being built. So, If I understand correctly, one could add regularization to a conv / fc layer as follows:
```
net = conv_2d(input, 96, 11, strides=4, activation='relu', regularizer='L2', decay=0.001)
net = ...
:
net = fully_connected(network, 4096, regularizer='L2', decay=0.001)
```
I assume that somehow TFlearn collects all of the defined regularized weights and combines them into the loss function. Is that correct?

I think there may be a benefit in adding regularization as I suggested above: it allows you to regularize all weights and avoid specifying the _regularizer_ parameter for each and every conv / fc layer. I still wonder if this is the way to do it.
